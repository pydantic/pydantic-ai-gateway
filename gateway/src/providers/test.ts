import { DefaultProviderProxy } from './default'

export class TestProvider extends DefaultProviderProxy {
  providerId(): string {
    return 'openai'
  }

  apiFlavour(): string | undefined {
    return 'chat'
  }

  fetch(url: string): Promise<Response> {
    const data = {
      choices: [
        {
          finish_reason: 'stop',
          index: 0,
          message: { annotations: [], content: `request URL: ${url}`, refusal: null, role: 'assistant' },
        },
      ],
      created: Math.floor(Date.now() / 1000),
      id: 'chatcmpl-test',
      model: 'gpt-5',
      object: 'chat.completion',
      service_tier: 'default',
      system_fingerprint: null,
      usage: {
        prompt_tokens: 4560,
        completion_tokens: 1230,
        completion_tokens_details: {
          accepted_prediction_tokens: 0,
          audio_tokens: 0,
          reasoning_tokens: 0,
          rejected_prediction_tokens: 0,
        },
        prompt_tokens_details: { audio_tokens: 0, cached_tokens: 0 },
        total_tokens: 5790,
      },
    }
    const headers = { 'Content-Type': 'application/json', 'pydantic-ai-gateway': 'test' }
    return Promise.resolve(new Response(JSON.stringify(data), { status: 200, headers }))
  }
}
