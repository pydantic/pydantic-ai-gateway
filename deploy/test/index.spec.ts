import { env, fetchMock, SELF } from 'cloudflare:test'
import OpenAI from 'openai'
import { afterEach, beforeAll, beforeEach, describe, expect, it } from 'vitest'
import SQL from '../../gateway/limits-schema.sql?raw'

beforeAll(async () => {
  try {
    const response = await fetch('http://localhost:8005')
    expect(response.status, 'The Proxy VCR seems to be facing issues, please check the logs.').toBe(204)
  } catch {
    throw new Error('Proxy VCR is not running. Run `make run-proxy-vcr` to enable tests.')
  }
})

const RESET_SQL = `\
DROP TABLE IF EXISTS spend;
DROP TABLE IF EXISTS keyStatus;

${SQL}`

beforeEach(async () => {
  const keys = await env.KV.list()
  if (keys.keys.length !== 0) {
    throw new Error('KV store is not empty before test.')
  }
  await env.limitsDB.prepare(RESET_SQL).run()
  fetchMock.activate()
})
afterEach(() => {
  fetchMock.assertNoPendingInterceptors()
})

describe('index', () => {
  it('responds with index html', async () => {
    const response = await SELF.fetch('https://example.com')
    expect(response.status).toBe(200)
  })
})

function recordOtelBatch(otelBatch: string[]) {
  fetchMock
    .get('https://logfire.pydantic.dev')
    .intercept({ method: 'POST', path: '/v1/traces', headers: { Authorization: 'write-token' } })
    .reply(({ body }) => {
      if (typeof body === 'string') {
        otelBatch.push(body)
      } else {
        throw new Error('Unexpected response body type')
      }
      return { statusCode: 200, body }
    })
}

describe('deploy', () => {
  it('status auth works', async () => {
    const noAuth = await SELF.fetch('https://example.com/status/')
    expect(noAuth.status).toBe(401)
    expect(await noAuth.text()).toMatchInlineSnapshot(`"Unauthorized - Missing "Authorization" Header"`)

    const wrongAuth = await SELF.fetch('https://example.com/status/', { headers: { authorization: 'wrong' } })
    expect(wrongAuth.status).toBe(401)
    expect(await wrongAuth.text()).toMatchInlineSnapshot(`"Unauthorized - Invalid API Key"`)

    const ok = await SELF.fetch('https://example.com/status/', { headers: { authorization: 'testing' } })
    expect(ok.status).toBe(200)
    expect(await ok.text()).toMatchSnapshot('status-empty')
  })

  it('should call openai via gateway', async () => {
    const otelBatch: string[] = []
    recordOtelBatch(otelBatch)

    const client = new OpenAI({
      apiKey: 'healthy-key',
      baseURL: 'https://example.com/openai',
      fetch: SELF.fetch.bind(SELF),
    })

    const completion = await client.chat.completions.create({
      model: 'gpt-5',
      messages: [
        { role: 'developer', content: 'You are a helpful assistant.' },
        { role: 'user', content: 'What is the capital of France?' },
      ],
    })

    expect(completion).toMatchSnapshot('llm')
    expect(otelBatch.length).toBe(1)
    expect(JSON.parse(otelBatch[0]!).resourceSpans?.[0].scopeSpans?.[0].spans?.[0]?.attributes).toMatchSnapshot('span')

    const response = await SELF.fetch('https://example.com/status/', { headers: { authorization: 'testing' } })
    expect(response.status).toBe(200)
    let data = await response.text()
    data = data
      .replace(/\d{4}-\d{2}-\d{2}/g, 'YYYY-MM-DD')
      .replace(/"raw": ?\d+/g, '"raw": 123456')
      .trim()
    expect(data).toMatchSnapshot('status-after-requests')
  })
})
