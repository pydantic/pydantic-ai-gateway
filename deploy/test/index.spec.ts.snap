// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`openai > should call openai via gateway > llm 1`] = `
{
  "choices": [
    {
      "finish_reason": "stop",
      "index": 0,
      "message": {
        "annotations": [],
        "content": "Paris.",
        "refusal": null,
        "role": "assistant",
      },
    },
  ],
  "created": 1757316780,
  "id": "chatcmpl-CDQeyDf30d6s2eFqr1PmAs2sdQcAd",
  "model": "gpt-5-2025-08-07",
  "object": "chat.completion",
  "service_tier": "default",
  "system_fingerprint": null,
  "usage": {
    "completion_tokens": 11,
    "completion_tokens_details": {
      "accepted_prediction_tokens": 0,
      "audio_tokens": 0,
      "reasoning_tokens": 0,
      "rejected_prediction_tokens": 0,
    },
    "prompt_tokens": 23,
    "prompt_tokens_details": {
      "audio_tokens": 0,
      "cached_tokens": 0,
    },
    "pydantic_ai_gateway": {
      "cost_estimate": 0.00013875,
    },
    "total_tokens": 34,
  },
}
`;

exports[`openai > should call openai via gateway > span 1`] = `
[
  {
    "key": "logfire.msg",
    "value": {
      "stringValue": "chat gpt-5-2025-08-07",
    },
  },
  {
    "key": "logfire.json_schema",
    "value": {
      "stringValue": "{"type":"object","properties":{"gen_ai.operation.name":{"type":"string"},"gen_ai.request.model":{"type":"string"},"gen_ai.system":{"type":"string"},"http.response.status_code":{"type":"number"},"http.request.body.text":{"type":"string"},"http.response.body.text":{"type":"string"},"gen_ai.response.model":{"type":"string"},"gen_ai.usage.input_tokens":{"type":"number"},"gen_ai.usage.cache_read_tokens":{"type":"number"},"gen_ai.usage.cache_write_tokens":{},"gen_ai.usage.output_tokens":{"type":"number"},"gen_ai.usage.input_audio_tokens":{},"gen_ai.usage.cache_audio_read_tokens":{},"gen_ai.usage.output_audio_tokens":{"type":"number"},"events":{}}}",
    },
  },
  {
    "key": "logfire.level_num",
    "value": {
      "intValue": 9,
    },
  },
  {
    "key": "gen_ai.operation.name",
    "value": {
      "stringValue": "chat",
    },
  },
  {
    "key": "gen_ai.request.model",
    "value": {
      "stringValue": "gpt-5",
    },
  },
  {
    "key": "gen_ai.system",
    "value": {
      "stringValue": "openai",
    },
  },
  {
    "key": "http.response.status_code",
    "value": {
      "intValue": 200,
    },
  },
  {
    "key": "http.request.body.text",
    "value": {
      "stringValue": "{
  "model": "gpt-5",
  "messages": [
    {
      "role": "developer",
      "content": "You are a helpful assistant."
    },
    {
      "role": "user",
      "content": "What is the capital of France?"
    }
  ]
}",
    },
  },
  {
    "key": "http.response.body.text",
    "value": {
      "stringValue": "{"id":"chatcmpl-CDQeyDf30d6s2eFqr1PmAs2sdQcAd","object":"chat.completion","created":1757316780,"model":"gpt-5-2025-08-07","choices":[{"index":0,"message":{"role":"assistant","content":"Paris.","refusal":null,"annotations":[]},"finish_reason":"stop"}],"usage":{"prompt_tokens":23,"completion_tokens":11,"total_tokens":34,"prompt_tokens_details":{"cached_tokens":0,"audio_tokens":0},"completion_tokens_details":{"reasoning_tokens":0,"audio_tokens":0,"accepted_prediction_tokens":0,"rejected_prediction_tokens":0},"pydantic_ai_gateway":{"cost_estimate":0.00013875}},"service_tier":"default","system_fingerprint":null}",
    },
  },
  {
    "key": "gen_ai.response.model",
    "value": {
      "stringValue": "gpt-5-2025-08-07",
    },
  },
  {
    "key": "gen_ai.usage.input_tokens",
    "value": {
      "intValue": 23,
    },
  },
  {
    "key": "gen_ai.usage.cache_read_tokens",
    "value": {
      "intValue": 0,
    },
  },
  {
    "key": "gen_ai.usage.cache_write_tokens",
    "value": {},
  },
  {
    "key": "gen_ai.usage.output_tokens",
    "value": {
      "intValue": 11,
    },
  },
  {
    "key": "gen_ai.usage.input_audio_tokens",
    "value": {},
  },
  {
    "key": "gen_ai.usage.cache_audio_read_tokens",
    "value": {},
  },
  {
    "key": "gen_ai.usage.output_audio_tokens",
    "value": {
      "intValue": 0,
    },
  },
  {
    "key": "events",
    "value": {
      "arrayValue": {
        "values": [],
      },
    },
  },
]
`;
